src/hyp_map.rs

```diff
@@ -209,6 +209,44 @@ impl PrivateRegion {
             }
         }
     }
+
+    // Restore private region to initial-state.
+    fn restore(&self) {
+        let mut copied = 0;
+        // We have to reset the full pages mapped for this segment.
+        let mapped_size = PageSize::Size4k.round_up(self.size as u64) as usize;
+        // Copy data at the beginning if it's present.
+        if let Some(data) = self.data {
+            // In case data is bigger than region size, write up to region end only.
+            let len = core::cmp::min(self.size, data.len());
+            let data = &data[0..len];
+            // Copy original data to umode area.
+            // Write to user mapping setting SUM in SSTATUS.
+            CSR.sstatus.modify(sstatus::sum.val(1));
+            // Safety:
+            // - this write is in a umode regions guaranteed to be mapped by HypMap in every page table.
+            // - the region starts at self.vaddr and is self.size byte long. `len` is <= `self.size`.
+            unsafe {
+                core::ptr::copy(data.as_ptr(), self.vaddr.bits() as *mut u8, len);
+            }
+            // Restore SUM.
+            CSR.sstatus.modify(sstatus::sum.val(0));
```

Comment:
Is SUM guaranteed to start at 0 when running `restore`?

Reply:
It's up to what we mean by guarantee, I suppose. Sstatus is set to zero in salus, and we only use it here. Every time we modify it (when we enter guest or u-mode) we save it and restore it.